src/pipeline_migration/actions/migrate.py

```diff
@@ -689,6 +703,145 @@ def _resolve_migrations(
+            prefix=groups["prefix"],
+            version=groups["version"],
+            file_checksum=groups["checksum"],
+            timestamp=groups["timestamp"],
+        )
+
+
+MIGRATION_IMAGE_TAG_REGEX: Final = re.compile(
+    "^"
+    r"(?P<prefix>migration)-"
+    r"(?P<version>\d+\.\d+(\.\d+)?)-"
+    r"(?P<checksum>[0-9a-f]{64})-"
+    r"(?P<timestamp>\d+)"
+    "$"
+)
+
+MIGRATION_IMAGE_TAG_LIKE_PATTERN: Final = r"migration-%.%-%-%"
+
+
+class MigrationImagesResolver(Resolver):
+
+    def _resolve_migrations(
+        self,
+        bundle_upgrade: TaskBundleUpgrade,
+        upgrades_range: list[QuayTagInfo],
+    ) -> Generator[TaskBundleMigration, Any, None]:
+        """Resolve migrations for a bundle upgrade
+
+        Migrations are pushed to the registry as OCI images tagged in a specific form:
+
+            migration-<actual task version>-<file sha256sum>-<timestamp>
+
+        * `migration-`: is a fixed prefix.
+        * `actual task version`: is the task version set in label ``app.kubernetes.io/version``.
+          They are treated as semantic versions.
+        * `file sha256sum`: is the checksum calculated in SHA256 algorithm.
+        * `timestamp`: is the time when a migration is pushed.
+
+        Every bundle has to be tagged with the actual task version. When Renovate sends an update
+        pull request, ``(current_value, new_value]`` will represent a straightforward upgrade range,
+        for example (0.2.1, 0.3.4], that are used to filter migrations.
+
+        :param bundle_upgrade: Refer to :meth:`Resolver._resolve_migrations`.
+        :param upgrades_range: Useless for this resolver. Any value passed-in is ignored.
+        :return: a generator yielding found migration represented by a TaskBundleMigration instance.
+            Yielded migrations are ensured to be in the correct order from oldest to newest version.
+        :raises ValueError: It is not allowed to modify an existing migration. The tekton bundle
+            build pipeline ensures such modification does not happen. But, in case a modified
+            migration is present in the image repository accidentally, resolver will stop proceeding
+            immediately.
+        :raises ValueError: If a migration image includes more than one layers (files), ValueError
+            will be raised.
+        """
+        old_version = parse_version(bundle_upgrade.current_value)
+        new_version = parse_version(bundle_upgrade.new_value)
+        if old_version == new_version:
+            # By tagging bundles with actual task version, e.g. 0.2, 0.2.1, 0.3.2, before bumping
+            # the actual task version, bundles within current version point to the migration that
+            # has been applied previously.
+            # So, in this in-version bundle update, there is no new migration to apply.
+            return
+        image_repo = bundle_upgrade.dep_name
+        migrations: list[tuple[str, TaskBundleMigration]] = []
```

Comment:
Maybe this should be a `list[tuple[Version, TaskBundleMigration]]`? Because we sort the list by version later, and I'm not sure if sorting by string is the same as sorting by semver